File: archived/src/c/ch01/1.2/l2/modexp.c
#include "modexp.h"

long long modexp(long long a, long long b, long long m) {
    long long result = 1;
    long long base = a % m;
    long long exp = b;

    while (exp > 0) {
        if (exp & 1) {
            result = (result * base) % m;
        }
        base = (base * base) % m;
        exp >>= 1;
    }
    return result;
}

File: archived/v1/src/erlang/ch01/1.2/src/modexp.erl
-module(modexp).
-export([modexp/3, demo/0]).

%% Modular exponentiation using square-and-multiply
modexp(A, B, M) ->
    modexp(A rem M, B, M, 1).

modexp(_, 0, _, Result) -> Result;
modexp(Base, Exp, M, Result) when (Exp band 1) =:= 1 ->
    modexp((Base*Base) rem M, Exp bsr 1, M, (Result*Base) rem M);
modexp(Base, Exp, M, Result) ->
    modexp((Base*Base) rem M, Exp bsr 1, M, Result).

demo() ->
    io:format("modexp(7,128,13) = ~p~n", [modexp(7,128,13)]).
